pythonwhat/State.py

```diff
@@ -169,13 +169,25 @@ def has_different_processes(self):
             # play it safe (most common)
             return True
 
+    # todo: remove (deprecated)
     def assert_root(self, fun, extra_msg=""):
         if self.parent_state is not None:
             raise InstructorError(
                 "`%s()` should only be called from the root state, `Ex()`. %s"
                 % (fun, extra_msg)
             )
 
+    def assert_execution_root(self, fun, extra_msg=""):
+        if not (
+            self.parent_state is None
+            or self.creator
+            and self.creator.get("type") == "run"
+        ):
```

Comment:
Not super readable IMO. This might be better:

```python
if self.parent_state is not None
  and (
    self.creator is None
    or self.creator.get("type") != "run"
  ):
  # ...
```

Or better, create a function `has_run_creator` and use `not(self.has_run_creator(...))`.

Reply:
That's what I had before writing this I think 😅

Reply:
Clarified with a helper method